amp/sharer.go

```diff
@@ -0,0 +1,156 @@
+}
+
+// SeedSharer orchestrates the sharing of the root AMP seed along multiple
+// paths. It also suports derivation of the child payment hashes that get
+// attached to HTLCs, and the child preimages used by the receiver to settle
+// individual HTLCs in the set.
+type SeedSharer struct {
+	root Share
+	curr Share
+}
+
+// NewSeedSharer generates a new SeedSharer instance with a seed drawn at
+// random.
+func NewSeedSharer() (*SeedSharer, error) {
+	var root Share
+	if _, err := rand.Read(root[:]); err != nil {
+		return nil, err
+	}
+
+	return SeedSharerFromRoot(&root), nil
+}
+
+// SeedSharerFromRoot instantiates a SeedSharer with an externally provided
+// seed.
+func SeedSharerFromRoot(root *Share) *SeedSharer {
+	return initSeedSharer(root, root)
+}
+
+func initSeedSharer(root, curr *Share) *SeedSharer {
+	return &SeedSharer{
+		root: *root,
+		curr: *curr,
+	}
+}
+
+// Seed returns the sharer's seed, the primary source of entropy for deriving
+// shares of the root.
+func (s *SeedSharer) Root() Share {
+	return s.root
+}
+
+// Share generates a new child of the root seed.
+//
+// NOTE: This method will fail after FinalShare has been called.
+func (s *SeedSharer) Split() (Sharer, Sharer, error) {
+	shareLeft, shareRight, err := split(&s.curr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	left := initSeedSharer(&s.root, &shareLeft)
+	right := initSeedSharer(&s.root, &shareRight)
+
+	return left, right, nil
+}
+
+// FinalShare ...
+//
+// NOTE: This method will fail after FinalShare has been called.
+func (s *SeedSharer) Child(index uint32) *Child {
+	desc := ChildDesc{
+		Share: s.curr,
+		Index: index,
+	}
+
+	return DeriveChild(s.root, desc)
+}
+
+// ReconstructChildren derives the set of children hashes and preimages from the
+// provided descriptors. The shares from each child descriptor are first used to
+// compute the root, afterwards the child hashes and preimages are
+// determinstically computed. For child descriptor at index i in the input, it's
+// derived child will occupy index i of the returned children.
+func ReconstructChildren(descs ...ChildDesc) []*Child {
+	// Recompute the root by XORing the provided shares.
+	var root Share
+	for _, desc := range descs {
+		xor(&root, &root, &desc.Share)
+	}
+
+	// With the root computed, derive the child hashes and preimages from
+	// the child descriptors.
+	children := make([]*Child, len(descs))
+	for i, desc := range descs {
+		children[i] = DeriveChild(root, desc)
+	}
+
+	return children
+}
+
+// split splits a share into two random values, that when XOR'd reproduce the
+// original share. Given a share s, the two shares are dervied as:
+//   left <-$- random
+//   right = parent ^ left.
+//
+// When reconstructed, we have that:
+//   left ^ right = left ^ parent ^ left
+//                = parent.
+func split(parent *Share) (Share, Share, error) {
+	// Generate a random share for the left child.
+	var left Share
+	if _, err := rand.Read(left[:]); err != nil {
```

Comment:
Just to double check, the rationale for not instead using the root share as the seed to a CSPRNG here is that we don't retry failed payments on start up, so it doesn't matter that these splits are deterministic?

Reply:
Yes, we currently don't resume MPP payments on restarts and instead just wait for them to be cancelled. The plan is to keep this for AMP as well, which means we don't need to be able to rederive any of the secret values.

FWIW the splits could still use an RNG if we choose to store them all, the deterministic derivation was just an attempt at minimizing the storage requirements and minimize how much randomness we consume.